### ISSUE TITLE
Standard Bidding Strategy Update in Existing Campaign

### ISSUE THREAD

**Issue by muthuqumar** [googleads/google-ads-java#524]
For updating a standard bidding strategy as part of the campaign object, we will need to set the new bidding strategy that is needed. For instance, an existing campaign has a manual_cpc bidding strategy and I am modifying the campaign object by setting the target spend. 

Since this is an existing campaign, we compare set the Update Mask (after referring the [guide](https://developers.google.com/google-ads/api/docs/client-libs/java/field-masks)) like below:
```
CampaignOperation campaignOperation =
    CampaignOperation.newBuilder()
        .setUpdate(campaignWrapper.getCampaign())
        .setUpdateMask(
          FieldMasks.compare(originalCampaign, modifiedCampaign)
        )
        .build();
```
And, this returns "manual_cpc" as well as "target_spend" since it compares with the existing campaign. So, running this operation leads to an error `FIELD_CANNOT_BE_CLEARED` for "manual_cpc". 

Also, the below utility gives us all the fields that are set as part of the current campaign object. 
`FieldMasks.allSetFieldsOf(campaignWrapper.getCampaign())`
So, according to the example I mentioned this returns "target_spend" but not "manual_cpc". 

As a workaround we solved this by taking the common ones from both the field masks, and set those fields as Update mask, like below:
```
List<String> allCampaignSetFields = FieldMasks.allSetFieldsOf(campaignWrapper.getCampaign())
    .getPathsList().stream().collect(toList());
List<String> modifedFields = FieldMasks.compare(originalCampaignWrapper.getCampaign(), campaignWrapper.getCampaign())
    .getPathsList().stream().collect(Collectors.toList());
allCampaignSetFields.retainAll(modifedFields);
```
Hence, our campaign operation is changed to:
```
CampaignOperation campaignOperation =
    CampaignOperation.newBuilder()
        .setUpdate(campaignWrapper.getCampaign())
        .setUpdateMask(
          FieldMask.newBuilder().addAllPaths(campaignSetFields)
        )
        .build();
```
It will be nice to have a separate method for FieldMask to compare original and modified campaign objects, as this case seems very specific to bidding strategy and not any other objects by far.

**Comment by nwbirnie** [comment]
I think we should probably remove the FieldMasks utility. There's numerous error cases such as this. We also don't really have a good way to implement this correctly and consistently for the whole API. @jradcliff @AnashOommen thoughts?

**Comment by PeterLavetsky** [comment]
I would ask to please not remove the FieldMask utility without some drop in replacement.

It is as the core of our integration facade on top of the Ads API, heavily relying on the FieldMasks.allSetFieldsOf() method.

At this point in time we are absolutely looking for increased stability from the Ads API after taking nearly a year to migrate from AdWords API to Ads API and now facing another tectonic migration in the form of Assets.

**Comment by devchas** [comment]
This is a bit of a quirk with the FieldMask utility in how it deals with oneof fields, but I don't think it warrants the removal of the FieldMask utility, which is rather useful. There is a separate [issue](https://github.com/googleads/google-ads-java/issues/272) related to FieldMasks and updating the campaign bidding strategy, but that issue is more so related to the core API functionality than the FieldMasks utility. (I'll follow up on that thread with more detail).

When you use the `FieldMasks.compare()` method, it adds all fields that are different in the two objects to the UpdateMask. Because `campaign_bidding_strategy` is a oneof field, you can only change it by setting a different bidding strategy, which will inherently "clear" the existing one. However, when both bidding strategies (the empty version of the old one, since it's not in the new campaign, and the new one) are present in the UpdateMask, the API sees two types of updates to perform on the campaign: clear the `manual_cpc` field and update the `target_spend` field, and the former is not allowed. You cannot clear a bidding strategy, only change it. 

The easiest solution for this is to use the `FieldMasks.allsetFieldsOf()` method instead of `FieldMasks.compare()` when updating bidding_strategies (or other oneof fields that cannot be cleared). This way, the empty `manual_cpc` message isn't included in the UpdateMask.

That begs the question, what if you want to compare two campaigns (campaign1 and campaign2), see which fields are have changed from campaign1 -> campaign2, and only update the fields that have changed. In that case you could still use `FieldMasks.allsetFieldsOf()` on campaign2. 

The only case where that might be an issue is if you are removing a field from campaign1 because it is not present in campaign2. Say, for example, campaign1 had the `name` field set, but campaign2 did not. If your intention is to remove the campaign `name` from the campaign, then you would need to use `FieldMasks.compare()` because `FieldMasks.allsetFieldsOf()` would not pick that up on campaign2, as there is the `name` field is not present.

However, if you are concerned that your UpdateMask will include fields that have NOT changed, than this really isn't an issue so long as the fields in the campaign2 object are the same as in campaign1. Say for example, the `name` of the campaign has NOT been updated, but it IS included on campaign2. In this case, using `FieldMasks.allsetFieldsOf()` will result in `name` being included in the UpdateMask. However, since the `name` in both campaigns is the same, there will be no impact on the campaign itself. The only downside is that the payload may be unnecessarily larger, but it will not impact the outcome of the operation. 

**To summarize, use the `FieldMasks.allsetFieldsOf()` method instead of `FieldMasks.compare()` with the caveat being that `FieldMasks.allsetFieldsOf()` will not pick up on fields that have been removed (and that you want to remove) in the modified object.**

**Comment by devchas** [comment]
@muthuqumar I am going to close out this issue because the `FieldMasks` utility is actually working as intended, and using the `FieldMasks.allsetFieldsOf()` method, as described above, should solve your issue. However, please feel free to open this back up if you have additional questions or if you feel that the approach I described does not work for your use case.
